src/action/list_directory.rs

```diff
@@ -0,0 +1,689 @@
+    symlink: Option<PathBuf>,
+    path: PathBuf,
+    crtime: Option<SystemTime>,
+}
+
+impl Default for Response {
+
+    fn default() -> Response {
+        Response {
+            mode: None,
+            ino: None,
+            dev: None,
+            nlink: None,
+            uid: None,
+            gid: None,
+            size: None,
+            atime: None,
+            mtime: None,
+            ctime: None,
+            blocks: None,
+            blksize: None,
+            rdev: None,
+            flags_linux: None,
+            symlink: None,
+            path: Default::default(),
+            crtime: None,
+        }
+    }
+}
+
+/// A request type for the list directory action.
+pub struct Request {
+    path: PathBuf,
+}
+
+fn get_accesses_time(metadata: &Metadata) -> Option<SystemTime> {
+    match metadata.accessed() {
+        Ok(atime) => Some(atime),
+        Err(err) => {
+            warn!("unable to get last access time: {}", err);
+            None
+        }
+    }
+}
+
+fn get_modification_time(metadata: &Metadata) -> Option<SystemTime> {
+    match metadata.modified() {
+        Ok(atime) => Some(atime),
+        Err(err) => {
+            warn!("unable to get last modification time: {}", err);
+            None
+        }
+    }
+}
+
+fn get_creation_time(metadata: &Metadata) -> Option<SystemTime> {
+    match metadata.created() {
+        Ok(atime) => Some(atime),
+        Err(err) => {
+            warn!("unable to get creation time: {}", err);
+            None
+        }
+    }
+}
+
+#[cfg(target_os = "linux")]
+fn get_status_change_time(metadata: &Metadata) -> Option<SystemTime> {
+    use std::time::Duration;
+    use std::os::unix::fs::MetadataExt;
+
+    UNIX_EPOCH.checked_add(Duration::from_secs(metadata.ctime() as u64))
+}
+
+
+#[cfg(target_os = "linux")]
+fn fill_response(metadata: &Metadata, file_path: &Path) -> Response {
+    use std::os::unix::fs::MetadataExt;
+
+    Response {
+        mode: Some(metadata.mode().into()),
+        ino: Some(metadata.ino() as u32),
+        dev: Some(metadata.dev() as u32),
+        nlink: Some(metadata.nlink() as u32),
+        uid: Some(metadata.uid() as u32),
+        gid: Some(metadata.gid() as u32),
+        size: Some(metadata.size()),
+        atime: get_accesses_time(&metadata),
+        mtime: get_modification_time(&metadata),
+        ctime: get_status_change_time(&metadata),
+        blocks: Some(metadata.blocks() as u32),
+        blksize: Some(metadata.blksize() as u32),
+        rdev: Some(metadata.rdev() as u32),
+        flags_linux:
+        Some(get_linux_flags(file_path).unwrap_or_default() as u32),
+        symlink: if metadata.file_type().is_symlink() {
+            match fs::read_link(file_path) {
+                Ok(file) => Some(file),
+                Err(error) => {
+                    warn!("unable to read symlink: {}", error);
+                    None
+                }
+            }
+        } else {
+            None
+        },
```

Comment:
This has some non-trivial logic and does I/O, it should not be inside struct initialization.

Reply:
I've put this code into a separate function.